File: src/lib/waterfall.ts
import type { NodeStatus, EdgeType } from '@/types/graph';

export interface WaterfallNode {
  id: string;
  status: NodeStatus;
  parentId: string | null;
}

export interface WaterfallEdge {
  sourceId: string;
  targetId: string;
  edgeType: EdgeType;
}

export interface StatusChange {
  nodeId: string;
  oldStatus: NodeStatus;
  newStatus: NodeStatus;
}

const MAX_DEPTH = 10;

/**
 * Compute cascading status changes triggered by a node's status update.
 *
 * Pure function — does not touch the DB.  The caller is responsible for
 * persisting the returned changes and recording history.
 */
export function computeWaterfall(
  changedNodeId: string,
  newStatus: NodeStatus,
  allNodes: WaterfallNode[],
  allEdges: WaterfallEdge[],
): StatusChange[] {
  // Only PASS and FAIL trigger propagation
  if (newStatus !== 'PASS' && newStatus !== 'FAIL') return [];

  const nodeMap = new Map(allNodes.map((n) => [n.id, { ...n }]));
  // Apply the triggering change first
  const trigger = nodeMap.get(changedNodeId);
  if (trigger) trigger.status = newStatus;

  const changes: StatusChange[] = [];
  const visited = new Set<string>([changedNodeId]);

  // BFS queue: [nodeId, depth]
  const queue: [string, number][] = [[changedNodeId, 0]];

  while (queue.length > 0) {
    const [currentId, depth] = queue.shift()!;
    if (depth >= MAX_DEPTH) continue;

    const current = nodeMap.get(currentId);
    if (!current) continue;

    const derived = deriveChanges(currentId, current.status, nodeMap, allEdges);

    for (const change of derived) {
      if (visited.has(change.nodeId)) continue;
      visited.add(change.nodeId);

      const target = nodeMap.get(change.nodeId);
      if (!target) continue;
      if (target.status === change.newStatus) continue;

      change.oldStatus = target.status;
      target.status = change.newStatus;
      changes.push(change);
      queue.push([change.nodeId, depth + 1]);
    }

    // Parent-child aggregation: if current node has a parent, re-evaluate the parent
    if (current.parentId && !visited.has(current.parentId)) {
      const parentChange = evaluateParent(current.parentId, nodeMap, allNodes);
      if (parentChange) {
        visited.add(parentChange.nodeId);
        const parent = nodeMap.get(parentChange.nodeId);
        if (parent) {
          parent.status = parentChange.newStatus;
          changes.push(parentChange);
          queue.push([parentChange.nodeId, depth + 1]);
        }
      }
    }
  }

  return changes;
}

/** Derive direct status changes from a single node's new status. */
function deriveChanges(
  nodeId: string,
  status: NodeStatus,
  nodeMap: Map<string, WaterfallNode>,
  allEdges: WaterfallEdge[],
): StatusChange[] {
  const results: StatusChange[] = [];

  for (const edge of allEdges) {
    if (edge.edgeType === 'related' || edge.edgeType === 'parent_child') continue;

    // ── blocks: A --blocks--> B ──
    if (edge.edgeType === 'blocks' && edge.sourceId === nodeId) {
      const target = nodeMap.get(edge.targetId);
      if (!target) continue;

      if (status === 'PASS' && target.status === 'BLOCKED') {
        // Check all other blockers of this target
        const otherBlockers = allEdges.filter(
          (e) => e.edgeType === 'blocks' && e.targetId === edge.targetId && e.sourceId !== nodeId,
        );
        const allResolved = otherBlockers.every((e) => nodeMap.get(e.sourceId)?.status === 'PASS');
        if (allResolved) {
          results.push({ nodeId: edge.targetId, oldStatus: target.status, newStatus: 'IN_PROGRESS' });
        }
      } else if (status === 'FAIL' && target.status === 'IN_PROGRESS') {
        results.push({ nodeId: edge.targetId, oldStatus: target.status, newStatus: 'BLOCKED' });
      }
    }

    // ── depends_on: A --depends_on--> B  (A depends on B) ──
    if (edge.edgeType === 'depends_on' && edge.targetId === nodeId) {
      const dependent = nodeMap.get(edge.sourceId);
      if (!dependent) continue;

      if (status === 'PASS' && dependent.status === 'BLOCKED') {
        // Check all dependencies of this dependent
        const allDeps = allEdges.filter(
          (e) => e.edgeType === 'depends_on' && e.sourceId === edge.sourceId,
        );
        const allMet = allDeps.every((e) => nodeMap.get(e.targetId)?.status === 'PASS');
        if (allMet) {
          results.push({ nodeId: edge.sourceId, oldStatus: dependent.status, newStatus: 'IN_PROGRESS' });
        }
      } else if (status === 'FAIL' && dependent.status === 'IN_PROGRESS') {
        results.push({ nodeId: edge.sourceId, oldStatus: dependent.status, newStatus: 'BLOCKED' });
      }
    }

    // ── triggers: A --triggers--> B ──
    if (edge.edgeType === 'triggers' && edge.sourceId === nodeId) {
      const target = nodeMap.get(edge.targetId);
      if (!target) continue;

      if (status === 'PASS' && target.status === 'BLOCKED') {
        results.push({ nodeId: edge.targetId, oldStatus: target.status, newStatus: 'IN_PROGRESS' });
      }
    }
  }

  return results;
}

/** Evaluate GROUP parent status from children. */
function evaluateParent(
  parentId: string,
  nodeMap: Map<string, WaterfallNode>,
  allNodes: WaterfallNode[],
): StatusChange | null {
  const parent = nodeMap.get(parentId);
  if (!parent) return null;

  const children = allNodes.filter((n) => n.parentId === parentId);
  if (children.length === 0) return null;

  const statuses = children.map((c) => nodeMap.get(c.id)?.status ?? c.status);

  let newStatus: NodeStatus;
  if (statuses.every((s) => s === 'PASS')) {
    newStatus = 'PASS';
  } else if (statuses.some((s) => s === 'FAIL')) {
    newStatus = 'FAIL';
  } else {
    newStatus = 'IN_PROGRESS';
  }

  if (newStatus === parent.status) return null;

  return { nodeId: parentId, oldStatus: parent.status, newStatus };
}

File: backend/internal/service/waterfall.go
package service

import "github.com/thask/backend/internal/model"

const maxWaterfallDepth = 10

type StatusChange struct {
	NodeID    string           `json:"nodeId"`
	OldStatus model.NodeStatus `json:"oldStatus"`
	NewStatus model.NodeStatus `json:"newStatus"`
}

type WaterfallNode struct {
	ID       string
	Status   model.NodeStatus
	ParentID *string
}

type WaterfallEdge struct {
	SourceID string
	TargetID string
	EdgeType model.EdgeType
}

// ComputeWaterfall calculates cascading status changes triggered by a node's status update.
// Pure function — does not touch the DB.
func ComputeWaterfall(changedNodeID string, newStatus model.NodeStatus, allNodes []WaterfallNode, allEdges []WaterfallEdge) []StatusChange {
	if newStatus != model.NodeStatusPass && newStatus != model.NodeStatusFail {
		return nil
	}

	nodeMap := make(map[string]*WaterfallNode, len(allNodes))
	for i := range allNodes {
		n := allNodes[i]
		nodeMap[n.ID] = &WaterfallNode{ID: n.ID, Status: n.Status, ParentID: n.ParentID}
	}

	// Apply the triggering change
	if trigger, ok := nodeMap[changedNodeID]; ok {
		trigger.Status = newStatus
	}

	var changes []StatusChange
	visited := map[string]bool{changedNodeID: true}

	type queueItem struct {
		id    string
		depth int
	}
	queue := []queueItem{{changedNodeID, 0}}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.depth >= maxWaterfallDepth {
			continue
		}

		node, ok := nodeMap[current.id]
		if !ok {
			continue
		}

		derived := deriveChanges(current.id, node.Status, nodeMap, allEdges)
		for _, change := range derived {
			if visited[change.NodeID] {
				continue
			}
			visited[change.NodeID] = true

			target, ok := nodeMap[change.NodeID]
			if !ok || target.Status == change.NewStatus {
				continue
			}

			change.OldStatus = target.Status
			target.Status = change.NewStatus
			changes = append(changes, change)
			queue = append(queue, queueItem{change.NodeID, current.depth + 1})
		}

		// Parent-child aggregation
		if node.ParentID != nil && !visited[*node.ParentID] {
			if parentChange := evaluateParent(*node.ParentID, nodeMap, allNodes); parentChange != nil {
				visited[parentChange.NodeID] = true
				if parent, ok := nodeMap[parentChange.NodeID]; ok {
					parent.Status = parentChange.NewStatus
					changes = append(changes, *parentChange)
					queue = append(queue, queueItem{parentChange.NodeID, current.depth + 1})
				}
			}
		}
	}

	return changes
}

func deriveChanges(nodeID string, status model.NodeStatus, nodeMap map[string]*WaterfallNode, allEdges []WaterfallEdge) []StatusChange {
	var results []StatusChange

	for _, edge := range allEdges {
		if edge.EdgeType == model.EdgeTypeRelated || edge.EdgeType == model.EdgeTypeParentChild {
			continue
		}

		// blocks: A --blocks--> B
		if edge.EdgeType == model.EdgeTypeBlocks && edge.SourceID == nodeID {
			target, ok := nodeMap[edge.TargetID]
			if !ok {
				continue
			}
			if status == model.NodeStatusPass && target.Status == model.NodeStatusBlocked {
				allResolved := true
				for _, e := range allEdges {
					if e.EdgeType == model.EdgeTypeBlocks && e.TargetID == edge.TargetID && e.SourceID != nodeID {
						if src, ok := nodeMap[e.SourceID]; ok && src.Status != model.NodeStatusPass {
							allResolved = false
							break
						}
					}
				}
				if allResolved {
					results = append(results, StatusChange{NodeID: edge.TargetID, OldStatus: target.Status, NewStatus: model.NodeStatusInProgress})
				}
			} else if status == model.NodeStatusFail && target.Status == model.NodeStatusInProgress {
				results = append(results, StatusChange{NodeID: edge.TargetID, OldStatus: target.Status, NewStatus: model.NodeStatusBlocked})
			}
		}

		// depends_on: A --depends_on--> B (A depends on B)
		if edge.EdgeType == model.EdgeTypeDependsOn && edge.TargetID == nodeID {
			dependent, ok := nodeMap[edge.SourceID]
			if !ok {
				continue
			}
			if status == model.NodeStatusPass && dependent.Status == model.NodeStatusBlocked {
				allMet := true
				for _, e := range allEdges {
					if e.EdgeType == model.EdgeTypeDependsOn && e.SourceID == edge.SourceID {
						if dep, ok := nodeMap[e.TargetID]; ok && dep.Status != model.NodeStatusPass {
							allMet = false
							break
						}
					}
				}
				if allMet {
					results = append(results, StatusChange{NodeID: edge.SourceID, OldStatus: dependent.Status, NewStatus: model.NodeStatusInProgress})
				}
			} else if status == model.NodeStatusFail && dependent.Status == model.NodeStatusInProgress {
				results = append(results, StatusChange{NodeID: edge.SourceID, OldStatus: dependent.Status, NewStatus: model.NodeStatusBlocked})
			}
		}

		// triggers: A --triggers--> B
		if edge.EdgeType == model.EdgeTypeTriggers && edge.SourceID == nodeID {
			target, ok := nodeMap[edge.TargetID]
			if !ok {
				continue
			}
			if status == model.NodeStatusPass && target.Status == model.NodeStatusBlocked {
				results = append(results, StatusChange{NodeID: edge.TargetID, OldStatus: target.Status, NewStatus: model.NodeStatusInProgress})
			}
		}
	}

	return results
}

func evaluateParent(parentID string, nodeMap map[string]*WaterfallNode, allNodes []WaterfallNode) *StatusChange {
	parent, ok := nodeMap[parentID]
	if !ok {
		return nil
	}

	var children []WaterfallNode
	for _, n := range allNodes {
		if n.ParentID != nil && *n.ParentID == parentID {
			children = append(children, n)
		}
	}
	if len(children) == 0 {
		return nil
	}

	allPass := true
	anyFail := false
	for _, c := range children {
		s := c.Status
		if node, ok := nodeMap[c.ID]; ok {
			s = node.Status
		}
		if s != model.NodeStatusPass {
			allPass = false
		}
		if s == model.NodeStatusFail {
			anyFail = true
		}
	}

	var newStatus model.NodeStatus
	if allPass {
		newStatus = model.NodeStatusPass
	} else if anyFail {
		newStatus = model.NodeStatusFail
	} else {
		newStatus = model.NodeStatusInProgress
	}

	if newStatus == parent.Status {
		return nil
	}

	return &StatusChange{NodeID: parentID, OldStatus: parent.Status, NewStatus: newStatus}
}
